spacy/language.py

```diff
@@ -257,9 +270,28 @@ def factory_names(self) -> List[str]:
         """
         return list(self.factories.keys())
 
+    @property
+    def component_names(self) -> List[str]:
+        """Get the names of the available pipeline components. Includes all
+        active and inactive pipeline components.
+
+        RETURNS (List[str]): List of component name strings, in order.
+        """
+        return [pipe_name for pipe_name, _ in self.components]
+
+    @property
+    def pipeline(self) -> List[Tuple[str, Callable[[Doc], Doc]]]:
```

Comment:
I hadn't thought about the fact that we're returning lists for most of these properties. I think we should change them to tuples.

It's especially important for the pipeline, because users may well have code like `nlp.pipeline.append`. That will silently do nothing if we have this as a list.

Reply:
Okay, maybe we could have our own frozen list subclass, though (similar to `SimpleFrozenDict`) that lets us raise custom errors? The default error for a tuple is always kinda cryptic (especially if the user doesn't understand what's going on), so it'd be nice to raise more specific errors here.

Reply:
Sounds great, definitely up for that.

Reply:
(Make it a tuple subclass though?)

Reply:
I think we'd have to use some `__getattr__ trickery to do the raising though. We can't just have an `append` method, or it'll mess up people's type checks.

Reply:
Naming ideas:

* `ComputedTuple`
* `PropertyTuple`
* `ViewTuple`
* `AttributeView`
* `ImmutableView`